src/react/databrowser/objects/Objects.jsx

```diff
@@ -30,14 +33,40 @@ export default function Objects() {
   const listType = useSelector((state: AppState) => state.s3.listObjectsType);
   const bucketInfo = useSelector((state: AppState) => state.s3.bucketInfo);
 
-  const toggled = useMemo(() => objects.filter(o => o.toggled), [objects]);
+  const query = useQuery();
+  const isShowVersions = query.get('showversions') === 'true';
+  const searchInput = query.get('metadatasearch');
+  const objectKey = query.get('prefix');
+  const versionId = query.get('versionId');
+
+  const toggled = useMemo(
+    () =>
+      objects.filter(
+        o =>
+          o.toggled ||
+          (!isShowVersions && o.key === objectKey) ||
+          (isShowVersions && o.key === objectKey && o.versionId === versionId),
+      ),
+    [objects, isShowVersions, objectKey, versionId],
+  );
   const { bucketName: bucketNameParam } = useParams();
   const prefixWithSlash = usePrefixWithSlash();
+
   useEffect(() => {
-    dispatch(
-      listObjects(bucketNameParam, prefixWithSlash, LIST_OBJECTS_S3_TYPE),
-    ).finally(() => setLoaded(true));
-  }, [bucketNameParam, prefixWithSlash, dispatch]);
+    if (searchInput) {
+      dispatch(newSearchListing(bucketNameParam, searchInput)).finally(() =>
+        setLoaded(true),
+      );
+    } else {
+      dispatch(
+        listObjects(
+          bucketNameParam,
+          prefixWithSlash,
+          isShowVersions ? LIST_OBJECT_VERSIONS_S3_TYPE : LIST_OBJECTS_S3_TYPE,
+        ),
+      ).finally(() => setLoaded(true));
```

Comment:
i'm not sure the dispatch function work as a async function. Not sure that the finally will run after all actions are done.

Reply:
Since here I only have one action, it should be OK.